In [8]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.grid_search import GridSearchCV
from sklearn.cross_validation import train_test_split
from sklearn.metrics import roc_auc_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import roc_curve
from sklearn import cross_validation
import math

%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [9]:
data = pd.read_csv("breast_cancer.csv")

In [10]:
data = data.drop(['Unnamed: 0', 'id number'], axis=1)
y = data.pop("malignant")

In [11]:
X_train, X_test, y_train, y_test = train_test_split(data, y, test_size=.2, random_state=42)

In [12]:
n_estimators = [300,400,500]
max_features = ['auto', 'sqrt','log2']
min_samples_split = [3,5,7]


rfc = RandomForestClassifier(n_jobs=1)
estimator = GridSearchCV(rfc,
                         dict(n_estimators=n_estimators,
                              max_features=max_features,
                              min_samples_split=min_samples_split
                             ), cv=None, n_jobs=1, refit=True)
                              

In [23]:
#estimator.fit(X_train, y_train)
print ("Overwritten for Readability")

Overwritten for Readability


In [14]:
estimator.best_estimator_

C:\Users\Graham\Anaconda3\lib\site-packages\sklearn\base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='sqrt', max_leaf_nodes=None,
            min_samples_leaf=1, min_samples_split=5,
            min_weight_fraction_leaf=0.0, n_estimators=300, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [15]:
best_rfc = estimator.best_estimator_

### Accuracy



In [16]:
accuracy = accuracy_score(y_test, best_rfc.predict(X_test))
print("Accuracy: ", accuracy)

Accuracy:  0.964285714286


### Precision and Recall

In [17]:
print (classification_report(y_test, best_rfc.predict(X_test)))

             precision    recall  f1-score   support

          0       0.98      0.97      0.97        95
          1       0.93      0.96      0.95        45

avg / total       0.96      0.96      0.96       140



### AUC

In [18]:
roc = roc_auc_score(y_test, best_rfc.predict_proba(X_test)[:,1])
print ("AUC Score: ", roc)

AUC Score:  0.996257309942


In [21]:
#scores = cross_validation.cross_val_score(best_rfc, data, y, cv=10)
print ("Overwritten for Readability")

Overwritten for Readability


In [20]:
mean_score = scores.mean()
std_dev = scores.std()
std_error = scores.std() / math.sqrt(scores.shape[0])
ci =  2.262 * std_error
lower_bound = mean_score - ci
upper_bound = mean_score + ci

print ("Score is %f +/-  %f" % (mean_score, ci))
print ('95 percent probability that if this experiment were repeated over and over the average score would be between %f and %f' % (lower_bound, upper_bound))

Score is 0.970080 +/-  0.018412
95 percent probability that if this experiment were repeated over and over the average score would be between 0.951668 and 0.988492


### Summary

The AUC of the model was 0.996 (using a holdout)

The precision and recall are both 0.96

The validation showed an overfitting of the data, as the range for the AUC should fall between 0.951668 and 0.988492 (with an average of 0.97)